### PR TITLE
Added ShouldBeJsonSerializable

### PR DESCRIPTION
**Pull request by rikrak**
Added ShouldBeJsonSerializable (see #3 )
Note: I've added the extension to the root namespace (__FluentAssertions__) to aid discovery.

This is based on a previous PR[1] by @borismod that was never merged

[1] https://github.com/fluentassertions/fluentassertions.json/commit/54432fdebc39e8ac52f5c9a0a30c08bdab01bbcf

### CODE REVIEW

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                    .IncludingProperties();
+
+                var typedSubject = (T)assertions.Subject;
+                ((T)deserializedObject).Should().BeEquivalentTo(typedSubject, _ => options(defaultOptions));
+            }
+            catch (Exception exc)
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith("Expected {0} to be JSON serializable{reason}, but serialization failed with:\r\n\r\n{1}",
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:82]: I recommend rewriting the failure as

`FailWith("Expected {object:context} to be JSON serializable{reason}, but serializing {0} failed with {1}", assertions.Subject, exc)`

This will make sure that the name of the variable on which you call `Should()` will be included in the message. But you need to add `[CustomAssertion]` to all your new methods. And by passing in the entire exception object, you let FA figure out how to format that the best way.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs (hunk 1 above)
Comment (rikrak) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs]: Thanks!  I've updated the PR as suggested

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            return new AndConstraint<ObjectAssertions>(assertions);
+        }
+
+        private static object CreateCloneUsingJsonSerializer(object subject)
+        {
+            var serializedObject = JsonConvert.SerializeObject(subject);
+            var cloneUsingJsonSerializer = JsonConvert.DeserializeObject(serializedObject, subject.GetType());
+            return cloneUsingJsonSerializer; ;
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:92]: 🙃 Nitpick: unnecessary semi-colon

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            try
+            {
+                var deserializedObject = CreateCloneUsingJsonSerializer(assertions.Subject);
+
+                var defaultOptions = AssertionOptions.CloneDefaults<T>()
+                    .RespectingRuntimeTypes()
+                    .IncludingFields()
+                    .IncludingProperties();
+
+                var typedSubject = (T)assertions.Subject;
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:75]: 🤔 What if the subject is not of type `T`?
🤔 What if the subject is `null`?

File: Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        public void Class_does_not_have_default_constructor()
+        {
+            // arrange
+            var target = _fixture.Create<PocoWithNoDefaultConstructor>();
+
+            // act
+            Action act = () => target.Should().BeJsonSerializable();
+
+            // assert
+            act.Should().Throw<Xunit.Sdk.XunitException>();
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:57]: 🤔 I think we should also verify a part of the message to make sure the test fails for the right reasons. Same for the other tests where you expect an exception.

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using FluentAssertions.Equivalency;
+using FluentAssertions.Execution;
+using FluentAssertions.Primitives;
+using Newtonsoft.Json;
+
+// added to the root namespace to aid discovery
+// ReSharper disable CheckNamespace
+namespace FluentAssertions
```
Comment (dennisdoomen) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:9]: ❌ Please use the same namespace as the rest of the code base.

File: Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        private readonly Fixture _fixture;
+
+        public ShouldBeJsonSerializableTests()
+        {
+            _fixture = new Fixture();
+        }
+
+        [Fact]
+        public void Class_is_simple_poco_should_be_serializable()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:22]: 🤔 I don't quite get this naming convention in all of these test cases. For this one, did you intend to say "Class as simple poco should be serializable"? Maybe "Simple_poco_should_be_serializable" or "When_class_is_simple_poco_it_should_be_serializable"?

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Apologies, but I noticed I wasn't finished with the review.

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs (hunk 3 above)
Comment (rikrak) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:75]: With respect to what to do if the subject is `null`: Both `BeBinarySerializable()` and `BeDataContractSerializable()` fail the assertion with what looks like a `NullReferenceException`. I could do the same, but I was thinking of something more like this:
```
    Execute.Assertion.ForCondition(assertions.Subject != null)
        .BecauseOf(because, becauseArgs)
        .FailWith("Expected {object:context} to be JSON serializable{reason}, but the value is null.  Please provide a value for the assertion.");
```
Not quite sure on the fail message, but it gets the point across I think

Comment (dennisdoomen) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:75]: Yeah, that's sloppy old code. The main library will do exactly what you suggested.

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs (hunk 5 above)
Comment (rikrak) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs]: Fixed in https://github.com/fluentassertions/fluentassertions.json/pull/52/commits/c4ee603f6e7a2ec108af29611af0fd1f715c6715

**Comment by rikrak** [comment]
I'm not too familiar with the etiquette of github - Should I mark the review comments as _Resolved_ after a commit, or is that up to the reviewer?  Either way I've just (hopefully!) addressed the review comments with the latest commits - let me know if they're OK, or if there are any other bits to look at :-)

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs (hunk 3 above)
Comment (rikrak) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:83]: Done :-)

**Review by dennisdoomen — APPROVED** [review]

**Comment by dennisdoomen** [comment]
Thank you for your contribution.

**Comment by dennisdoomen** [comment]
Crap. Completely forgot to ask @jnyrup to also take a peek...

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace FluentAssertions.Json
+{
+    /// <summary>
+    ///     Contains extension methods for JSON serialization assertion methods
+    /// </summary>
+    public static class ObjectAssertionsExtensions
+    {
+        /// <summary>
+        /// Asserts that an object can be serialized and deserialized using the JSON serializer and that it stills retains
```
Comment (jnyrup) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:15]: "stills retains" -> "still retains"

File: Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+            // act
+            Action act = () => target.Should().BeJsonSerializable();
+
+            // assert
+            act.Should()
+                .Throw<Xunit.Sdk.XunitException>(because:"This is consistent with BeBinarySerializable() and BeDataContractSerializable()")
+                .Which.Message
+                    .Should().Contain("value is null")
+                        .And.Contain("Please provide a value for the assertion");
```
Comment (jnyrup) [inline Tests/FluentAssertions.Json.Specs/ShouldBeJsonSerializableTests.cs:113]: Tip, it should be possible to shorten this to:
```cs
.WithMessage("*value is null*Please provide a value for the assertion*");
```

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+            return new AndConstraint<ObjectAssertions>(assertions);
+        }
+
+        private static object CreateCloneUsingJsonSerializer(object subject)
+        {
+            var serializedObject = JsonConvert.SerializeObject(subject);
+            var cloneUsingJsonSerializer = JsonConvert.DeserializeObject(serializedObject, subject.GetType());
+            return cloneUsingJsonSerializer;
+        }
```
Comment (jnyrup) [inline Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:101]: Does it make any difference if we use `subject.GetType()` or `typeof(T)`?
E.g. for inheritance and interfaces.

**Review by jnyrup — COMMENTED** [review]
I was thinking out giving this a view later today.

Aside from the single question about inheritance, it looks good to me.

File: Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs (hunk 9 above)
Comment (rikrak) [reply Src/FluentAssertions.Json/ObjectAssertionsExtensions.cs:101]: Looks like using typeof(T) causes issues with the non-generic overload of BeJsonSerialisable as `T` is `object` which causes problems with the equivalency of the deserialised object.  Essentially a dynamic object is deserialised that doesn't play well with the IsEquivalentTo test.  In addition the deserialised dynamic object has only string properties rather than, for example, Guids or DateTimes that the original instance had.
